src/Pact/Eval.hs

```diff
@@ -521,42 +521,68 @@ typecheck ps = void $ foldM tvarCheck M.empty ps where
                         evalError (_tInfo t) $ "Type error: values for variable " ++ show _aType ++
                         " do not match: " ++ show (prevTy,ty)
 
-check1 :: forall e . Info -> Type (Term Name) -> Term Name -> Eval e (Maybe (TypeVar (Term Name),Type (Term Name)))
-check1 i spec t = do
+-- | 'typecheckTerm i spec t' checks a Term 't' against a specified type 'spec'.
+-- Returns `Nothing` on successful check against concrete/untyped,
+-- or `Just` a pair for successful check against a type variable, where
+-- the pair is the type variable itself and the term type.
+typecheckTerm :: forall e . Info -> Type (Term Name) -> Term Name
+       -> Eval e (Maybe (TypeVar (Term Name),Type (Term Name)))
+typecheckTerm i spec t = do
+
   ty <- case typeof t of
     Left s -> evalError i $ "Invalid type in value location: " ++ unpack s
     Right r -> return r
+
   let
+
     tcFail :: Show a => a -> Eval e b
-    tcFail found = evalError i $ "Type error: expected " ++ show spec ++ ", found " ++ show found
+    tcFail found = evalError i $
+      "Type error: expected " ++ show spec ++ ", found " ++ show found
+
     tcOK = return Nothing
-    paramCheck :: Eq t => Type t -> Type t -> (Type t -> Eval e (Type (Term Name))) ->
-                  Eval e (Maybe (TypeVar (Term Name),Type (Term Name)))
-    paramCheck TyAny _ _ = tcOK
-    paramCheck TyVar {} _ _ = tcOK
-    paramCheck pspec pty check | pspec == pty = tcOK -- dupe check as below, for totality
-                               | not (isUnconstrainedTy pty) = tcFail ty -- unequal constrained fails
```

Comment:
Another strange check and probably unreachable -- value types inferred to a constrained type variable would fail.